Authorize users from s3

ktp242 opened the pull request:
This PR adds the check to see if the logged in user authorized. To achieve this, we need three steps, 

1. Decoded the access token that we have gotten from NYPL SSO. The access token is a JWT encoded string, so we will use JWT module to decoded it. See line 118 - 125 of `app/controllers/oauth_controller.rb`

2. Get the authorized user list from our AWS S3. The list consists of a series of authorized email addresses. See line 127 -139 of `app/controllers/oauth_controller.rb`

3. Compare to see if the user's email exists on the list

We check the authorization right after we got the access token. If the user is authorized, proceed to redirect the user to the request route. If not, send the user to the error page. See line 60 - 65 of `app/controllers/oauth_controller.rb`

As of now, the error page is only contain one line of "Something is not right". In my next PR, I will build the page with correct error message and disable the navigation.

ktp242 committed:
Added the method to check if the logged in user authorized.

ktp242 committed:
Added the region by env variable.

ktp242 commented on app/controllers/oauth_controller.rb:
```diff
@@ -129,7 +129,7 @@ def is_user_authorized
         access_key_id: ENV['AWS_KEY_ID'],
         secret_access_key: ENV['AWS_SECRET'],
-        region: 'us-east-1'
+        region: ENV['SQS_REGION']
```
We might want to consider rename this env variable, if we use it more than just for SQS.

ktp242 committed:
Corrected the typo.

ktp242 commented:
Following the suggestion from @emu47 He and I will adjust this PR to build a user model that includes the email address. So we can also use the same model to construct the request to SQS.

ktp242 committed:
Updated the user as a model.

ktp242 commented:
I have update user as a model. The model has the method to get the logged in user's email address.

emu47 committed:
[KK] Moving methods from controller to user model

emu47 committed:
[KK] Some deletions I missed.

emu47 committed:
[KK] Adding user email to messages to sqs

ktp242 committed:
Added user_spec.rb for testing.

ktp242 committed:
Merge branch 'authorize-users-from-s3' of https://github.com/NYPL/scsbuster into authorize-users-from-s3

ktp242 committed:
Removed unnecessary importing of aws-sdk-s3 gem.

ktp242 committed:
Updated the mock tokens.

ktp242 commented:
@nodanaonlyzuul I have updated the PR based on Kris's suggestions and also, Kris has merged his changes to this PR. What this PR now does are

1. check to see if the logged in user authorized <= please see the description of this PR for details
2. add the user model so we can use the info of the logged in user -now only the email- to check if s/he is an authorized user and also send the email address to the SQS service. Please check oauth_controller.rb and items_controller.rb for the applications for these two methods.
3. Add a unit test, user_spec.rb for the newly built user model.

nodanaonlyzuul commented:
Hey @ktp242.

Can you see if you can move `app/controllers/concerns/error_controller.rb` up a directory to `app/controllers/error_controller.rb`? Also - you probably don't need to have the empty `def error` controller in there.

nodanaonlyzuul commented on app/models/user.rb:
```diff
@@ -0,0 +1,43 @@
+  def self.get_authorized_list
+    authorized_list = nil
+    begin
+      s3 = Aws::S3::Client.new({
```
Hey @ktp242, 

### Roles

Creating the S3 Client like this will only work in non-production....

In production we want to instantiate it without a key and secret since it's going to get its authorization/powers via the role of the server.

We do a similar thing in the [SqsClient](https://github.com/NYPL/scsbuster/blob/master/lib/sqs_client.rb#L4).
**Let me know if you want to chat about what IAM roles are.**

### Region

I feel like you can hard-wire the region here rather than reusing the `SQS_REGION` environment variable.

ktp242 replied:
Cool I can take a look on how we did that for SQS Client.

About the region, if we made the other region for SQS a env variable, I think maybe we should make it as well?

ktp242 committed:
Moved error_controller.rb to the right directory.

ktp242 committed:
Removed the error_controller.rb that is located in the wrong folder.

ktp242 commented:
Thanks I have moved error_controller.rb to the correct location.

ktp242 committed:
Updated the ways to construct the S3 client instance.

ktp242 committed:
Renamed the env variable from 'SQS_REGION' to 'AWS_REGION'.

nodanaonlyzuul commented on app/models/user.rb:
```diff
@@ -0,0 +1,51 @@
+    email_address
+  end
+  
+  def is_authorized?
```
Can this whole method be shortened to:

```ruby
authorized_list = User.get_authorized_list
authorized_list && authorized_list.include?(self.get_email_address)
```

nodanaonlyzuul committed:
Shorten logic of User#is_authorized?

nodanaonlyzuul committed:
Use the _path helper to redirect to error page

nodanaonlyzuul committed:
Define often-used mocks in a before() setup method

nodanaonlyzuul committed:
Make the authorized users bucket name an environment variable

This allows different environments (development, qa, production) to use
different authorization.json files.
